### PR TITLE
Adding Create and Delete SubmodelElement API

### PR DESCRIPTION
**Pull request by zhangzai123**

### CODE REVIEW

**Review by FrankSchnicke — CHANGES_REQUESTED** [review]
There are no tests for the new methods in SubmodelRepository, please add them. Additionally, SubmodelElementList & Entity are not covered by the implementation & tests.

File: basyx.http/src/test/java/org/eclipse/digitaltwin/basyx/http/serialization/BaSyxHttpTestUtils.java [hunk 1]
```diff
@@ -202,4 +187,10 @@ private static HttpGet createGetRequestWithHeader(String url) {
 		aasCreateRequest.setHeader("Accept", "application/json");
 		return aasCreateRequest;
 	}
+
+	private static HttpDelete createDeleteRequestHeader(String url) {
```
[the remark this answers is not in the document]
Comment (zhangzai123) [reply basyx.http/src/test/java/org/eclipse/digitaltwin/basyx/http/serialization/BaSyxHttpTestUtils.java:191]: checked

File: basyx.submodelrepository-core/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/SubmodelRepository.java [hunk 2]
```diff
@@ -129,4 +129,36 @@ public interface SubmodelRepository {
...
+	 * 
+	 * @param submodelId
+	 *            the Submodel id
+	 * @param smElement
+	 *            the SubmodelElement
+	 */
+	public void createSubmodelElement(String submodelId, SubmodelElement smElement);
+
+	/**
+	 * Creates a nested SubmodelElement in a submodelelement collection
```
[the remark this answers is not in the document]
Comment (zhangzai123) [reply basyx.submodelrepository-core/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/SubmodelRepository.java:144]: checked

File: basyx.submodelrepository-http/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/http/SubmodelRepositoryApiHTTPController.java [hunk 3]
```diff
@@ -179,4 +205,8 @@ private boolean isValueContentRequest(String content) {
 	private boolean isNormalContentRequest(String content) {
 		return content.equals("normal");
 	}
+
```
[the remark this answers is not in the document]
Comment (zhangzai123) [reply basyx.submodelrepository-http/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/http/SubmodelRepositoryApiHTTPController.java:208]: checked

File: basyx.submodelservice-core/src/main/java/org/eclipse/digitaltwin/basyx/submodelservice/SubmodelService.java [hunk 4]
```diff
@@ -76,10 +78,42 @@ public interface SubmodelService {
...
 
+	/**
+	 * Creates a Submodel Element
+	 * 
+	 * @return
+	 */
+	public void createSubmodelElement(SubmodelElement submodelElement);
+
+	/**
+	 * Create a submodel element in a submodel element collection
```
[the remark this answers is not in the document]
Comment (zhangzai123) [reply basyx.submodelservice-core/src/main/java/org/eclipse/digitaltwin/basyx/submodelservice/SubmodelService.java:98]: checked

File: basyx.submodelservice-core/src/main/java/org/eclipse/digitaltwin/basyx/submodelservice/SubmodelService.java [hunk 5]
```diff
@@ -76,10 +78,42 @@ public interface SubmodelService {
...
+	 * 
+	 * @param idShortPath
+	 *            the SubmodelElement IdShortPath
+	 * @param submodelElement
+	 *            the submodel element to be created
+	 * @throws ElementDoesNotExistException
+	 *             If the submodel element collection does not exist
```
[the remark this answers is not in the document]
Comment (zhangzai123) [reply basyx.submodelservice-core/src/main/java/org/eclipse/digitaltwin/basyx/submodelservice/SubmodelService.java:105]: checked

File: basyx.submodelservice-core/src/main/java/org/eclipse/digitaltwin/basyx/submodelservice/SubmodelService.java [hunk 6]
```diff
@@ -76,10 +78,42 @@ public interface SubmodelService {
...
+	 */
+	public void createSubmodelElement(String idShortPath, SubmodelElement submodelElement) throws ElementDoesNotExistException;
+
+	/**
+	 * Delete a submodel element in a submodel
+	 * 
+	 * @param idshortPath
+	 *            the SubmodelElement IdShortPath
+	 * @throws ElementDoesNotExistException
+	 *             If the submodel element collection does not exist
```
[the remark this answers is not in the document]
Comment (zhangzai123) [reply basyx.submodelservice-core/src/main/java/org/eclipse/digitaltwin/basyx/submodelservice/SubmodelService.java:115]: checked

File: basyx.submodelservice-core/src/main/java/org/eclipse/digitaltwin/basyx/submodelservice/pathParsing/HierarchicalSubmodelElementParser.java [hunk 7]
```diff
@@ -55,7 +56,8 @@ public HierarchicalSubmodelElementParser(Submodel submodel) {
 	/**
 	 * Returns the nested SubmodelElement given in the idShortPath
 	 * 
-	 * @param idShortPath the isShortPath of the SubmodelElement
+	 * @param idShortPath
+	 *            the isShortPath of the SubmodelElement
```
[the remark this answers is not in the document]
Comment (zhangzai123) [reply basyx.submodelservice-core/src/main/java/org/eclipse/digitaltwin/basyx/submodelservice/pathParsing/HierarchicalSubmodelElementParser.java:60]: checked

File: basyx.submodelservice-core/src/main/java/org/eclipse/digitaltwin/basyx/submodelservice/pathParsing/HierarchicalSubmodelElementParser.java [hunk 8]
```diff
@@ -67,6 +69,30 @@ public SubmodelElement getSubmodelElementFromIdShortPath(String idShortPath) thr
 		return getLastElementOfStack(idShortPathTokenStack);
 	}
 
+	/**
+	 * Check whether the given idShortPath is a simple siDhort or a path of idShorts
```
[the remark this answers is not in the document]
Comment (zhangzai123) [reply basyx.submodelservice-core/src/main/java/org/eclipse/digitaltwin/basyx/submodelservice/pathParsing/HierarchicalSubmodelElementParser.java:73]: checked

File: basyx.submodelservice-core/src/main/java/org/eclipse/digitaltwin/basyx/submodelservice/pathParsing/HierarchicalSubmodelElementParser.java [hunk 9]
```diff
@@ -67,6 +69,30 @@ public SubmodelElement getSubmodelElementFromIdShortPath(String idShortPath) thr
...
+	 * 
+	 * @param idShortPath
+	 * @return true if the given idShortPath is a path of IdShorts
+	 */
+	public boolean isNestedIdShortPath(String idShortPath) {
+		return idShortPath.contains(".");
```
[the remark this answers is not in the document]
Comment (zhangzai123) [reply basyx.submodelservice-core/src/main/java/org/eclipse/digitaltwin/basyx/submodelservice/pathParsing/HierarchicalSubmodelElementParser.java:79]: checked

File: basyx.submodelservice-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/submodelservice/InMemorySubmodelService.java [hunk 10]
```diff
@@ -75,16 +78,79 @@ public SubmodelElement getSubmodelElement(String idShortPath) throws ElementDoes
...
-		valueMapper.setValue(value);	
+
+		valueMapper.setValue(value);
+	}
+
+	@Override
+	public void createSubmodelElement(SubmodelElement submodelElement) throws CollidingIdentifierException {
+		List<SubmodelElement> smElements = submodel.getSubmodelElements();
+
+		if (submodelElementExistAlready(submodelElement, smElements)) {
```
[the remark this answers is not in the document]
Comment (zhangzai123) [reply basyx.submodelservice-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/submodelservice/InMemorySubmodelService.java:98]: checked

File: basyx.submodelservice-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/submodelservice/InMemorySubmodelService.java [hunk 11]
```diff
@@ -75,16 +78,79 @@ public SubmodelElement getSubmodelElement(String idShortPath) throws ElementDoes
...
+	private boolean submodelElementExistAlready(SubmodelElement submodelElement, Collection<SubmodelElement> smElements) {
+		List<SubmodelElement> sme = smElements.stream().filter(s -> s.getIdShort().equals(submodelElement.getIdShort())).collect(Collectors.toList());
+		return !sme.isEmpty();
+	}
+
+	@Override
+	public void createSubmodelElement(String idShortPath, SubmodelElement submodelElement) throws ElementDoesNotExistException, CollidingIdentifierException {
+		SubmodelElementCollection collection = (SubmodelElementCollection) parser.getSubmodelElementFromIdShortPath(idShortPath);
+		Collection<SubmodelElement> submodelElements = collection.getValue();
+		if (submodelElementExistAlready(submodelElement, submodelElements)) {
```
[the remark this answers is not in the document]
Comment (zhangzai123) [reply basyx.submodelservice-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/submodelservice/InMemorySubmodelService.java:115]: cehcked

File: basyx.submodelservice-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/submodelservice/InMemorySubmodelService.java [hunk 12]
```diff
@@ -75,16 +78,79 @@ public SubmodelElement getSubmodelElement(String idShortPath) throws ElementDoes
...
+
+		if (submodelElementExistAlready(submodelElement, smElements)) {
+			throw new CollidingIdentifierException(submodelElement.getIdShort());
+		}
+
+		smElements.add(submodelElement);
+		submodel.setSubmodelElements(smElements);
+	}
+
+	private boolean submodelElementExistAlready(SubmodelElement submodelElement, Collection<SubmodelElement> smElements) {
```
[the remark this answers is not in the document]
Comment (zhangzai123) [reply basyx.submodelservice-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/submodelservice/InMemorySubmodelService.java:106]: checked

File: basyx.submodelservice-core/src/test/java/org/eclipse/digitaltwin/basyx/submodelservice/SubmodelServiceSuite.java [hunk 13]
```diff
@@ -272,20 +269,79 @@ public void getFileValue() {
...
+
+		createNestedSubmodelElement(submodelService);
+
+		Property property = new DefaultProperty.Builder().kind(ModelingKind.INSTANCE).idShort("test456").category("cat1").value("305").valueType(DataTypeDefXsd.INTEGER).build();
+
+		String idShortPath = DummySubmodelFactory.SUBMODEL_ELEMENT_COLLECTION_TOP + "." + DummySubmodelFactory.SUBMODEL_ELEMENT_FIRST_LIST + "[0][0]";
+		submodelService.createSubmodelElement(idShortPath, property);
+		idShortPath = idShortPath + ".test456";
+		SubmodelElement created = submodelService.getSubmodelElement(idShortPath);
+		PropertyValue createdValue = (PropertyValue) submodelService.getSubmodelElementValue(idShortPath);
```
[the remark this answers is not in the document]
Comment (zhangzai123) [reply basyx.submodelservice-core/src/test/java/org/eclipse/digitaltwin/basyx/submodelservice/SubmodelServiceSuite.java:285]: checked

File: basyx.submodelservice-core/src/test/java/org/eclipse/digitaltwin/basyx/submodelservice/SubmodelServiceSuite.java [hunk 14]
```diff
@@ -272,20 +269,79 @@ public void getFileValue() {
 		assertEquals(expectedValue, submodelElementValue.getValue());
 	}
 
+	@Test
+	public void createNestedSubmodelElement() {
+		Submodel operationDataSubmodel = DummySubmodelFactory.createOperationalDataSubmodelWithHierarchicalSubmodelElements();
+		SubmodelService submodelService = getSubmodelService(operationDataSubmodel);
+
+		createNestedSubmodelElement(submodelService);
```
[the remark this answers is not in the document]
Comment (zhangzai123) [reply basyx.submodelservice-core/src/test/java/org/eclipse/digitaltwin/basyx/submodelservice/SubmodelServiceSuite.java:277]: checked

File: basyx.submodelservice-core/src/test/java/org/eclipse/digitaltwin/basyx/submodelservice/SubmodelServiceSuite.java [hunk 15]
```diff
@@ -272,20 +269,79 @@ public void getFileValue() {
...
+		PropertyValue sunmodelElementValue = (PropertyValue) submodelService.getSubmodelElementValue("test123");
+		assertEquals("test123", submodelEl.getIdShort());
+		assertEquals("205", sunmodelElementValue.getValue());
+
+		submodelService.deleteSubmodelElement("test123");
+		submodelService.getSubmodelElement("test123");
+	}
+
+	@Test(expected = ElementDoesNotExistException.class)
+	public void deleteNestedSubmodelElement() {
```
[the remark this answers is not in the document]
Comment (zhangzai123) [reply basyx.submodelservice-core/src/test/java/org/eclipse/digitaltwin/basyx/submodelservice/SubmodelServiceSuite.java:310]: checked

File: basyx.submodelservice-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/submodelservice/InMemorySubmodelService.java [hunk 16]
```diff
@@ -75,16 +78,79 @@ public SubmodelElement getSubmodelElement(String idShortPath) throws ElementDoes
...
+		List<SubmodelElement> sme = smElements.stream().filter(s -> s.getIdShort().equals(submodelElement.getIdShort())).collect(Collectors.toList());
+		return !sme.isEmpty();
+	}
+
+	@Override
+	public void createSubmodelElement(String idShortPath, SubmodelElement submodelElement) throws ElementDoesNotExistException, CollidingIdentifierException {
+		SubmodelElementCollection collection = (SubmodelElementCollection) parser.getSubmodelElementFromIdShortPath(idShortPath);
```
[the remark this answers is not in the document]
Comment (zhangzai123) [reply basyx.submodelservice-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/submodelservice/InMemorySubmodelService.java:113]: checked

File: basyx.submodelservice-core/src/main/java/org/eclipse/digitaltwin/basyx/submodelservice/pathParsing/HierarchicalSubmodelElementParser.java [hunk 17]
```diff
@@ -67,6 +69,30 @@ public SubmodelElement getSubmodelElementFromIdShortPath(String idShortPath) thr
 		return getLastElementOfStack(idShortPathTokenStack);
 	}
 
+	/**
+	 * Check whether the given idShortPath is a simple siDhort or a path of idShorts
+	 * 
+	 * @param idShortPath
+	 * @return true if the given idShortPath is a path of IdShorts
+	 */
+	public boolean isNestedIdShortPath(String idShortPath) {
```
[the remark this answers is not in the document]
Comment (zhangzai123) [reply basyx.submodelservice-core/src/main/java/org/eclipse/digitaltwin/basyx/submodelservice/pathParsing/HierarchicalSubmodelElementParser.java:78]: checked

File: basyx.submodelservice-core/src/main/java/org/eclipse/digitaltwin/basyx/submodelservice/pathParsing/HierarchicalSubmodelElementParser.java [hunk 18]
```diff
@@ -67,6 +69,30 @@ public SubmodelElement getSubmodelElementFromIdShortPath(String idShortPath) thr
...
+
+	/**
+	 * Extract the idShort of the direct parent (submodel element collection) of the
+	 * nested submodel element
+	 * 
+	 * @param idShortPath
+	 *            - A path to the nested target submodel element
+	 * @return idShort of the direct parent of the nested submodel element
+	 */
+	public String extractDirectParentIdShortOfNestedSubmodelElement(String idShortPath) {
```
[the remark this answers is not in the document]
Comment (zhangzai123) [reply basyx.submodelservice-core/src/main/java/org/eclipse/digitaltwin/basyx/submodelservice/pathParsing/HierarchicalSubmodelElementParser.java:90]: checked

File: basyx.submodelrepository-http/src/test/java/org/eclipse/digitaltwin/basyx/submodelrepository/http/TestSubmodelRepositorySubmodelElementsHTTP.java [hunk 19]
```diff
@@ -343,6 +343,68 @@ private String wrapStringValue(String value) {
 		return "\"" + value + "\"";
 	}
 
+	@Test
+	public void createExistingSubmodelElement() throws IOException {
```
[the remark this answers is not in the document]
Comment (zhangzai123) [reply basyx.submodelrepository-http/src/test/java/org/eclipse/digitaltwin/basyx/submodelrepository/http/TestSubmodelRepositorySubmodelElementsHTTP.java:347]: checked

**Pull request closed without merging**